unittests/BuildSystem/BuildExecutionQueueTest.cpp

```diff
@@ -0,0 +1,108 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "llbuild/BuildSystem/BuildExecutionQueue.h"
+#include "llbuild/BuildSystem/ExternalCommand.h"
+
+#include "gtest/gtest.h"
+
+using namespace llbuild;
+using namespace llbuild::core;
+using namespace llbuild::buildsystem;
+
+namespace {
+class PhonyCommand : public ExternalCommand {
+  public:
+    using ExternalCommand::ExternalCommand;
+
+    virtual bool shouldShowStatus() override { return false; }
+
+    virtual void getShortDescription(SmallVectorImpl<char> &result) override {
+      llvm::raw_svector_ostream(result) << getName();
+    }
+
+    virtual void getVerboseDescription(SmallVectorImpl<char> &result) override {
+      llvm::raw_svector_ostream(result) << getName();
+    }
+
+    virtual bool executeExternalCommand(BuildSystemCommandInterface& bsci,
+        Task* task,
+        QueueJobContext* context) override {
+      return true;
+    }
+};
+  
+class SimpleBuildExecutionQueueDelegate : public BuildExecutionQueueDelegate {
+
+  public:
+    SimpleBuildExecutionQueueDelegate() { 
+    }
+
+    virtual void commandJobStarted(Command* command) override {
+    }
+
+    virtual void commandJobFinished(Command* command) override {
+    }
+
+    virtual void commandProcessStarted(Command* command,
+        ProcessHandle handle) override {
+    }
+
+    virtual void commandProcessHadError(Command* command, ProcessHandle handle,
+        const Twine& message) override {
+    }
+
+    virtual void commandProcessHadOutput(Command* command, ProcessHandle handle,
+        StringRef data) override {
+    }
+
+    virtual void commandProcessFinished(Command* command, ProcessHandle handle,
+        int exitStatus) override {
+    }
+};
+
+class BuildExecutionQueueTest : public testing::Test {
+  protected:
+    SimpleBuildExecutionQueueDelegate delegate;
+    std::shared_ptr<BuildExecutionQueue> queue;
+    std::unique_ptr<PhonyCommand> phony;
+    std::string outputString;
+    BuildExecutionQueueTest() {
+      queue = std::shared_ptr<BuildExecutionQueue>(createLaneBasedExecutionQueue(delegate, 1));
+      phony = llvm::make_unique<PhonyCommand>("Phony");
+    }
+
+    virtual ~BuildExecutionQueueTest() {
+      llvm::outs() << "Destroying test queue\n"<< outputString << "<<>>> END";
+    }
+
+
+    void executeProcessTester(std::string exe) {
+      auto fn = [this](QueueJobContext* context) {
+        queue->executeProcess(context, {"/Users/aciid/mycode/swift.org/llbuild/unittests/BuildSystem/Inputs/BuildExecutionQueueExecuteProcessInputs/simple-output"}, {}, 
+            [this](StringRef output){ llvm::outs() << "output" << output; this->outputString.append(output);},
+            [this](StringRef error){ llvm::outs() << "err: " << error; });
+      };
+      queue->addJob({ phony.get(), std::move(fn) });
```

Comment:
@ddunbar I am having memory management issues here. The fixture object is destroyed before the callbacks are executed. How to make sure that that BuildExecutionQueueTest is not deallocated until the callbacks are alive?